lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -4165,17 +4199,19 @@ static ConstraintResult visitInherited(
   };
 
   // Visit all of the inherited types.
-  auto typeDecl = decl.dyn_cast<TypeDecl *>();
-  auto extDecl = decl.dyn_cast<ExtensionDecl *>();
   ArrayRef<TypeLoc> inheritedTypes = typeDecl ? typeDecl->getInherited()
                                               : extDecl->getInherited();
   for (unsigned index : indices(inheritedTypes)) {
     Type inheritedType = typeDecl ? typeDecl->getInheritedType(index)
                                   : extDecl->getInheritedType(index);
     if (!inheritedType) continue;
 
+    // If the inherited type is 'Any', allow it to pass through verbatim, as
+    // GSB diagnoses redundant 'Any' constraints.
+    bool skipDecomposition = inheritedType->isEqual(ctx.TheAnyType);
```

Comment:
Could this check be moved into `visitInherited`? Doing it out here both requires adding the argument somewhat unnecessarily (since the function has all the info it needs to do the computation itself), and also I suspect (but am not sure of the exact parser behaviour) it might miss cases like `Any & P` where the `Any` gets found recursively.

Reply:
I purposefully did the check outside of `visitInherited` in order to miss cases like `Any & P`, partly in order to have parity with the behaviour for protocol composition constraints that appear in `where` clauses (which aren't decomposed until after the diagnostic), and partly because I think we should diagnose redundant `Any` members in protocol compositions as a part of a separate broader diagnostic ([SR-8082]) that warns on the usage of `Any & ...` anywhere, not just in constraints.

Does this sound good? If so, I'll expand on the comment to include the above rationale for performing the check outside of `visitInherited`, and add a test to make sure we ignore `Any & P` constraints for now.

Thanks for the review!

  [SR-8082]: https://bugs.swift.org/browse/SR-8082

Reply:
> I think we should diagnose redundant Any members in protocol compositions as a part of a separate broader diagnostic (SR-8082) that warns on the usage of Any & ... anywhere, not just in constraints.

Ah, yeah, that's a more general and better approach.

> Does this sound good? If so, I'll expand on the comment to include the above rationale for performing the check outside of visitInherited, and add a test to make sure we ignore Any & P constraints for now.

👍